### PR TITLE
feat: add test 6.2.47

### PR DESCRIPTION
**Pull request by peinjoh**
This PR:
* moves the "generate canonical filename" and "has canonical url" functions into the trait, including test coverage
* moves 6.2.11, which previously contained the functions, to use them
* adds test 6.2.47, which needed a similar function "get all canonical urls"

Test coverage looks fine IMO, but I would invite you to discuss edge cases if you find any during review

### CODE REVIEW

File: csaf-rs/src/csaf/traits/document_trait.rs [hunk 1]
```diff
@@ -69,6 +84,20 @@ pub trait DocumentTrait {
...
+    /// Returns the canonical URLs from this document's references.
+    fn get_canonical_urls(&self) -> Vec<&str> {
+        let expected_filename = self.get_tracking().get_canonical_filename();
+        canonical_url_candidates(self.get_references(), expected_filename.as_str()).collect()
+    }
+
+    /// Returns `true` if the document has at least one canonical URL.
+    fn has_canonical_url(&self) -> bool {
+        let expected_filename = self.get_tracking().get_canonical_filename();
+        canonical_url_candidates(self.get_references(), expected_filename.as_str())
```
Comment (tziemek) [inline csaf-rs/src/csaf/traits/document_trait.rs:96]: Why not use `get_canonical_urls` here? Allocation of Vec?

**Review by tziemek — APPROVED** [review]

File: csaf-rs/src/csaf/traits/document_trait.rs (hunk 1 above)
Comment (peinjoh) [reply csaf-rs/src/csaf/traits/document_trait.rs:96]: Yes, get_canonical_urls will collect the Iterator into a Vec and for that runs each element through the 2 filters.

has_canonical_urls will only run until the first element matches both filters, so ideally less then all elements are evaled, and will not alloc a Vec
